jerry-core/vm/opcodes.c

```diff
@@ -258,18 +261,123 @@ opfunc_for_in (ecma_value_t left_value, /**< left value */
 
 /**
  * 'VM_OC_APPEND_ARRAY' opcode handler, for setting array object properties
+ *
+ * @return ECMA_VALUE_ERROR - if the operation failed
+ *         ECMA_VALUE_EMPTY, otherwise
  */
-void JERRY_ATTR_NOINLINE
+ecma_value_t JERRY_ATTR_NOINLINE
 opfunc_append_array (ecma_value_t *stack_top_p, /**< current stack top */
-                     uint8_t values_length) /**< number of elements to set */
+                     uint16_t values_length) /**< number of elements to set
+                                              *   with potential OPFUNC_HAS_SPREAD_ELEMENT flag */
 {
+#if ENABLED (JERRY_ES2015)
+  bool has_spread_element = false;
+  if (values_length >= OPFUNC_HAS_SPREAD_ELEMENT)
+  {
+    has_spread_element = true;
+    values_length = (uint16_t) (values_length & ~OPFUNC_HAS_SPREAD_ELEMENT);
+    JERRY_ASSERT (values_length < 255);
+  }
+#endif /* ENABLED (JERRY_ES2015) */
+
   ecma_object_t *array_obj_p = ecma_get_object_from_value (stack_top_p[-1]);
 
   JERRY_ASSERT (ecma_get_object_type (array_obj_p) == ECMA_OBJECT_TYPE_ARRAY);
 
   ecma_extended_object_t *ext_array_obj_p = (ecma_extended_object_t *) array_obj_p;
   uint32_t old_length = ext_array_obj_p->u.array.length;
 
+#if ENABLED (JERRY_ES2015)
+  if (has_spread_element)
+  {
+    for (uint32_t i = 0, j = old_length; i < values_length; i++)
+    {
+      if (!ecma_is_value_array_hole (stack_top_p[i]))
+      {
+        ecma_value_t ret_value = ECMA_VALUE_ERROR;
+        ecma_value_t put_comp;
```

Comment:
This should be moved into the code part where the variable is truly used, no need to have this at the start.